gh-139871: Add `bytearray.take_bytes([n])` to efficiently extract `bytes` (#140128)
python/cpython: Doc/library/stdtypes.rst

      .. versionadded:: 3.14

   .. method:: take_bytes(n=None, /)

      Take the first *n* bytes as an immutable :class:`bytes`. Defaults to all
      bytes.

      If *n* is negative indexes from the end and takes the first :func:`len`
      minus *n* bytes. If *n* is out of bounds raises :exc:`IndexError`.

      Taking less than the full length will leave remaining bytes in the
      :class:`bytearray` which requires a copy. If the remaining bytes should be
      discarded use :func:`~bytearray.resize` or :keyword:`del` to truncate
      then :func:`~bytearray.take_bytes` without a size.

      .. impl-detail::

         CPython implements this as a zero-copy operation making it a very

>          efficient way to make a :class:`bytes` from a :class:`bytearray`.

Reviewer: Only when n is None, no?

Author: Yea, I've been struggling to find a concise wording "minimal copying" leaves a lot more wiggle room to me; can just add the caveat "when taking all bytes" (`ba.take_bytes(len(ba))` also doesn't copy)

      .. list-table:: Suggested Replacements
         :header-rows: 1

         * - Description
           - Old
           - New

         * - Return :class:`bytes` after working with :class:`bytearray`
           - .. code:: python

                  def read() -> bytes:
                     buffer = bytearray(1024)
                     ...
                     return bytes(buffer)
           - .. code:: python

                  def read() -> bytes:
                     buffer = bytearray(1024)
                     ...
                     return buffer.take_bytes()

         * - Empty a buffer getting the bytes
           - .. code:: python

                  buffer = bytearray(1024)
                  ...
                  data = bytes(buffer)
                  buffer.clear()
           - .. code:: python

                  buffer = bytearray(1024)
                  ...
                  data = buffer.take_bytes()

>                   assert len(buffer) == 0

Reviewer: I suggest removing these assertions. We can trust take_bytes() API, no? :-)

Author: Will drop; was trying to show "buffer is empty / in same state" while avoiding entering full doctest / REPL land. Given they're in a table of equivalents should be good enough

      .. versionadded:: 3.14

   .. method:: take_bytes(n=None, /)

      Take the first *n* bytes as an immutable :class:`bytes`. Defaults to all

>       bytes.

Reviewer 2: This should define what *take* means.
```suggestion
      Remove the first *n* bytes from the bytearray and return them as an immutable
      :class:`bytes`.
      By default (if *n* is ``None``), return all bytes and clear the bytearray.
```

      If *n* is negative indexes from the end and takes the first :func:`len`

>       plus *n* bytes. If *n* is out of bounds raises :exc:`IndexError`.

Reviewer 2: Use the imperative:
```suggestion
      If *n* is negative, index from the end and take the first :func:`len`
      plus *n* bytes. If *n* is out of bounds, raise :exc:`IndexError`.
```

>       Taking less than the full length will leave remaining bytes in the
>       :class:`bytearray` which requires a copy. If the remaining bytes should be
>       discarded use :func:`~bytearray.resize` or :keyword:`del` to truncate
>       then :func:`~bytearray.take_bytes` without a size.

Reviewer 2: I think these clauses need commas:
```suggestion
      Taking less than the full length will leave remaining bytes in the
      :class:`bytearray`, which requires a copy. If the remaining bytes should be
      discarded, use :func:`~bytearray.resize` or :keyword:`del` to truncate
      then :func:`~bytearray.take_bytes` without a size.
```

      .. versionadded:: 3.14

   .. method:: take_bytes(n=None, /)

      Take the first *n* bytes as an immutable :class:`bytes`. Defaults to all
      bytes.

      If *n* is negative indexes from the end and takes the first :func:`len`
      plus *n* bytes. If *n* is out of bounds raises :exc:`IndexError`.

      Taking less than the full length will leave remaining bytes in the
      :class:`bytearray` which requires a copy. If the remaining bytes should be
      discarded use :func:`~bytearray.resize` or :keyword:`del` to truncate
      then :func:`~bytearray.take_bytes` without a size.

      .. impl-detail::

         Taking all bytes is a zero-copy operation.

>       .. list-table:: Suggested Replacements

Reviewer 2: Would it make sense to put porting notes better in What's New, rather than in general documentation? In a few years, `take_bytes` won't be “new”.
The `versionadded` note could link to the 3.15 What's New.

Author: End state to me is these (hopefully) get incorporated into `flake8` / `ruff` as code improvement suggestions rather than more things to memorize. To that end maybe just `bytes(bytearray())` -> `.take_bytes()` in what's new would be enough and rely on external pieces to socialize more (ex. I'm planning to try and teach this a bit through blogs/talks if it ships).

I slightly prefer on `bytearray` so someone hand-optimizing `bytearray` code and browsing the page for alternatives will find them. The less efficient patterns are stable Python APIs to me so it's similar to https://docs.python.org/3/library/pathlib.html#corresponding-tools and applies both in `3.15` specifically and forseeable future versions.

Reviewer 2: Maybe the table can be improved here, but as it is it looks out of place to me.

> End state to me is these (hopefully) get incorporated into flake8 / ruff as code improvement suggestions rather than more things to memorize. 

Exactly -- and the more ephemeral-looking What's New seems better for that.

> someone hand-optimizing bytearray code and browsing the page for alternatives will find them.

That's why I suggested the link to What's New :)

> to me so it's similar to https://docs.python.org/3/library/pathlib.html#corresponding-tools 

IMO, that's a very different thing: those docs say “pathlib is not a drop-in replacement for [os.path](https://docs.python.org/3/library/os.path.html#module-os.path)”. It's a two-way mapping, not “old→new” porting suggestions.

Author: (Still looking at / experimenting with this one, changes for others comments pushed)

Author: Moved to whatsnew; not fully satisfied with my wording but also assuming will rework over time (ex. reference `stdlib` improvements as they get added)

Reviewer 2: One benefit of What's New is that there'll be a reorganization pass closer to the release :)
